File: ts/examples/tool-router/src/openai-agents.ts

import { Composio } from "@composio/core";
import { Agent, run, hostedMcpTool } from "@openai/agents";
import { createInterface } from "node:readline/promises";

const composio = await new Composio()
const {mcp} = await composio.create("default");

const agent = new Agent({
  name: "Personal Assistant",
  model: "gpt-5.1",
  instructions: "You are a helpful personal assistant. Use Composio tools to execute tasks.",
  tools: [
    hostedMcpTool({ serverLabel: "composio", serverUrl: mcp.url, headers: {
      'x-api-key': process.env.COMPOSIO_API_KEY!,
    } }),
  ],
});

const rl = createInterface({ input: process.stdin, output: process.stdout });

for (let input: any = await rl.question("You: "); input !== "exit"; ) {
  const result = await run(agent, input);
  console.log(`Agent: ${result.finalOutput}\n`);
  input = [...result.history, { role: "user" as const, content: await rl.question("You: ") }];
}

rl.close();

File: python/examples/tool_router/openai_agents.py
import os

from agents import Agent, HostedMCPTool, Runner

from composio import Composio

composio = Composio()
session = composio.create(
    # A user with a connected Gmail account (raises KeyError when unset)
    user_id=os.environ["COMPOSIO_EXAMPLES_USER_ID"],
    # mcp=True surfaces session.mcp on the returned type
    mcp=True,
)

print(session.mcp)

composio_mcp = HostedMCPTool(
    tool_config={
        "type": "mcp",
        "server_label": "tool_router",
        "server_url": session.mcp.url,
        "require_approval": "never",
        "headers": session.mcp.headers,  # type: ignore[typeddict-item]  # headers is typed Dict[str, Optional[str]]; the SDK always populates it with str values
    }
)

agent = Agent(
    name="My Agent",
    instructions="You are a helpful assistant that can use the tools provided to you.",
    tools=[composio_mcp],
)

result = Runner.run_sync(
    starting_agent=agent,
    input="Find my last email and summarize it.",
)

print(result.final_output)
